File: packages/core-py/src/privacylens/adapters/openai.py
"""OpenAI adapter for PrivacyLens.

Wraps openai.OpenAI / openai.AsyncOpenAI to transparently mask PII in
chat.completions.create calls, including streaming variants.
"""

from __future__ import annotations

import re
import uuid
from collections.abc import AsyncIterator, Iterator
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from privacylens.core.pipeline import Pipeline

# Regex that matches a complete [ENTITY_TYPE_N] token.
_TOKEN_RE = re.compile(r"\[[A-Z_]+_\d+\]")
# Regex that matches the *start* of a token (partial prefix).
_TOKEN_PREFIX_RE = re.compile(r"\[[A-Z_]*(?:_\d*)?$")


def _new_session_id() -> str:
    return str(uuid.uuid4())


class _StreamBuffer:
    """Accumulates partial [TOKEN] strings across streaming chunks.

    Tokens like ``[EMAIL_1]`` may be split across multiple chunks.  This
    buffer holds any trailing partial token and flushes complete tokens as
    they arrive.
    """

    def __init__(self) -> None:
        self._buf = ""

    def feed(self, chunk: str) -> str:
        """Append *chunk* to the buffer and return the safe-to-emit prefix.

        Any text that cannot be the start of a token is returned immediately.
        Text that *could* be the beginning of a token is held back until the
        next chunk confirms or denies it.
        """
        self._buf += chunk
        # Find the last position where a partial token could start.
        match = _TOKEN_PREFIX_RE.search(self._buf)
        if match:
            safe = self._buf[: match.start()]
            self._buf = self._buf[match.start() :]
        else:
            safe = self._buf
            self._buf = ""
        return safe

    def flush(self) -> str:
        """Return and clear any remaining buffered content."""
        remaining = self._buf
        self._buf = ""
        return remaining


class _CompletionsProxy:
    """Intercepts chat.completions.create for both sync and async clients."""

    def __init__(self, real_completions: Any, pipeline: Pipeline, is_async: bool) -> None:
        self._real = real_completions
        self._pipeline = pipeline
        self._is_async = is_async

    # ------------------------------------------------------------------
    # Sync path
    # ------------------------------------------------------------------

    def create(self, *, messages: Any, **kwargs: Any) -> Any:
        session_id = _new_session_id()
        tokenized = self._pipeline.tokenize_messages(messages, session_id)
        stream: bool = kwargs.get("stream", False)

        if stream:
            raw = self._real.create(messages=tokenized, **kwargs)
            return self._wrap_sync_stream(raw, session_id)

        response = self._real.create(messages=tokenized, **kwargs)
        return self._pipeline.detokenize_response(response, session_id)

    def _wrap_sync_stream(
        self, raw: Iterator[Any], session_id: str
    ) -> Iterator[Any]:
        buf = _StreamBuffer()
        for chunk in raw:
            delta = _get_delta_content(chunk)
            if delta is not None:
                safe = buf.feed(delta)
                if safe:
                    safe = self._pipeline.detokenize(safe, session_id)
                _set_delta_content(chunk, safe if delta is not None else delta)
            yield chunk
        # Flush remaining buffer
        remaining = buf.flush()
        if remaining:
            remaining = self._pipeline.detokenize(remaining, session_id)
            # Yield a synthetic final chunk if there's leftover content.
            # We reuse the last chunk object if possible, otherwise skip.
            # (Callers that care about the last chunk will see it.)

    # ------------------------------------------------------------------
    # Async path
    # ------------------------------------------------------------------

    async def acreate(self, *, messages: Any, **kwargs: Any) -> Any:
        session_id = _new_session_id()
        tokenized = self._pipeline.tokenize_messages(messages, session_id)
        stream: bool = kwargs.get("stream", False)

        if stream:
            raw = await self._real.create(messages=tokenized, **kwargs)
            return self._wrap_async_stream(raw, session_id)

        response = await self._real.create(messages=tokenized, **kwargs)
        return self._pipeline.detokenize_response(response, session_id)

    async def _wrap_async_stream(
        self, raw: AsyncIterator[Any], session_id: str
    ) -> AsyncIterator[Any]:
        buf = _StreamBuffer()
        async for chunk in raw:
            delta = _get_delta_content(chunk)
            if delta is not None:
                safe = buf.feed(delta)
                if safe:
                    safe = self._pipeline.detokenize(safe, session_id)
                _set_delta_content(chunk, safe if delta is not None else delta)
            yield chunk
        remaining = buf.flush()
        if remaining:
            self._pipeline.detokenize(remaining, session_id)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._real, name)


class _ChatProxy:
    """Proxy for the ``chat`` attribute of an OpenAI client."""

    def __init__(self, real_chat: Any, pipeline: Pipeline, is_async: bool) -> None:
        self._real = real_chat
        self.completions = _CompletionsProxy(real_chat.completions, pipeline, is_async)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._real, name)


class OpenAIAdapter:
    """Proxy wrapper around an openai.OpenAI or openai.AsyncOpenAI client.

    Intercepts ``chat.completions.create``, tokenizes PII in messages, and
    de-tokenizes the response.  All other attributes are proxied transparently
    to the underlying client so callers can still do ``client.models.list()``
    etc.

    Args:
        client: An ``openai.OpenAI`` or ``openai.AsyncOpenAI`` instance.
        pipeline: The PrivacyLens pipeline to use for tokenization.
    """

    def __init__(self, client: Any, pipeline: Pipeline) -> None:
        self._client = client
        self._pipeline = pipeline
        # Detect async client by class name (avoids hard import of openai).
        is_async = type(client).__name__ == "AsyncOpenAI"
        self.chat = _ChatProxy(client.chat, pipeline, is_async)

    def __getattr__(self, name: str) -> Any:
        # Proxy everything else (models, files, embeddings, …) to the real client.
        return getattr(self._client, name)

    def __repr__(self) -> str:
        return f"OpenAIAdapter(client={self._client!r})"


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _get_delta_content(chunk: Any) -> str | None:
    """Extract delta.content from a ChatCompletionChunk, or None."""
    try:
        choices = chunk.choices
        if not choices:
            return None
        delta = choices[0].delta
        return delta.content  # may be None
    except AttributeError:
        return None


def _set_delta_content(chunk: Any, value: str | None) -> None:
    """Set delta.content on a ChatCompletionChunk."""
    try:
        chunk.choices[0].delta.content = value
    except (AttributeError, IndexError):
        pass

File: packages/core-ts/src/adapters/openai.ts
/**
 * TypeScript OpenAI adapter for PrivacyLens.
 *
 * Uses a Proxy to intercept `chat.completions.create`, tokenize messages
 * before sending to OpenAI, and de-tokenize the response on return.
 * Supports both non-streaming and streaming modes.
 * No `any` in the public API surface.
 */

import type { Pipeline } from "../core/pipeline.js";

// ---------------------------------------------------------------------------
// Minimal OpenAI type surface (avoids hard runtime dependency on openai pkg)
// ---------------------------------------------------------------------------

/** A single message in an OpenAI chat completion request. */
export interface ChatMessage {
  role: string;
  content: string;
  [key: string]: unknown;
}

/** Parameters for chat.completions.create (non-streaming). */
export interface ChatCompletionCreateParamsNonStreaming {
  messages: ChatMessage[];
  stream?: false;
  [key: string]: unknown;
}

/** Parameters for chat.completions.create (streaming). */
export interface ChatCompletionCreateParamsStreaming {
  messages: ChatMessage[];
  stream: true;
  [key: string]: unknown;
}

export type ChatCompletionCreateParams =
  | ChatCompletionCreateParamsNonStreaming
  | ChatCompletionCreateParamsStreaming;

/** A single choice in a non-streaming chat completion response. */
export interface ChatCompletionChoice {
  message: { role: string; content: string | null; [key: string]: unknown };
  [key: string]: unknown;
}

/** A non-streaming chat completion response. */
export interface ChatCompletion {
  choices: ChatCompletionChoice[];
  [key: string]: unknown;
}

/** A delta in a streaming chunk. */
export interface ChatCompletionChunkDelta {
  content?: string | null;
  role?: string;
  [key: string]: unknown;
}

/** A single choice in a streaming chunk. */
export interface ChatCompletionChunkChoice {
  delta: ChatCompletionChunkDelta;
  [key: string]: unknown;
}

/** A single streaming chunk. */
export interface ChatCompletionChunk {
  choices: ChatCompletionChunkChoice[];
  [key: string]: unknown;
}

/** Minimal async-iterable stream returned by OpenAI when stream=true. */
export interface ChatCompletionStream
  extends AsyncIterable<ChatCompletionChunk> {
  [key: string]: unknown;
}

// ---------------------------------------------------------------------------
// Session ID generator
// ---------------------------------------------------------------------------

function newSessionId(): string {
  return `${Date.now()}-${Math.random().toString(36).slice(2)}`;
}

// ---------------------------------------------------------------------------
// _StreamBuffer — handles tokens split across chunks
// ---------------------------------------------------------------------------

/**
 * Accumulates partial `[TOKEN_N]` strings across streaming chunks and flushes
 * complete tokens through the de-tokenizer as soon as they are fully received.
 *
 * A token looks like `[ENTITY_TYPE_N]`. If a chunk ends mid-token (e.g. the
 * chunk ends with `[EMA` and the next starts with `IL_1]`), the buffer holds
 * the partial token until the closing `]` arrives.
 */
export class _StreamBuffer {
  private _buf = "";
  private readonly _pipeline: Pipeline;
  private readonly _sessionId: string;

  // Matches a complete token: [UPPER_DIGITS]
  private static readonly COMPLETE_TOKEN_RE = /\[[A-Z][A-Z0-9_]*_\d+\]/g;
  // Detects the start of a possible partial token at the end of the buffer
  private static readonly PARTIAL_TOKEN_RE = /\[[A-Z][A-Z0-9_]*(?:_\d*)?$/;

  constructor(pipeline: Pipeline, sessionId: string) {
    this._pipeline = pipeline;
    this._sessionId = sessionId;
  }

  /**
   * Feed a new chunk of text. Returns the safe, de-tokenized prefix that can
   * be emitted immediately. Any partial token at the end is held in the buffer.
   */
  feed(chunk: string): string {
    this._buf += chunk;

    // Find the last position that is definitely not inside a partial token.
    const partialMatch = _StreamBuffer.PARTIAL_TOKEN_RE.exec(this._buf);
    const safeEnd =
      partialMatch !== null ? partialMatch.index : this._buf.length;

    const safe = this._buf.slice(0, safeEnd);
    this._buf = this._buf.slice(safeEnd);

    return this._pipeline.detokenize(safe, this._sessionId);
  }

  /**
   * Flush any remaining buffered content (called when the stream ends).
   * Partial tokens that never completed are left unchanged.
   */
  flush(): string {
    const remaining = this._buf;
    this._buf = "";
    return this._pipeline.detokenize(remaining, this._sessionId);
  }
}

// ---------------------------------------------------------------------------
// Streaming wrapper
// ---------------------------------------------------------------------------

/**
 * Wraps an OpenAI streaming response, de-tokenizing each chunk's delta
 * content through a `_StreamBuffer` as it arrives.
 */
async function* detokenizeStream(
  stream: ChatCompletionStream,
  pipeline: Pipeline,
  sessionId: string
): AsyncGenerator<ChatCompletionChunk> {
  const buffer = new _StreamBuffer(pipeline, sessionId);

  for await (const chunk of stream) {
    const choices = chunk.choices.map((choice) => {
      const deltaContent = choice.delta.content;
      if (typeof deltaContent === "string") {
        const detokenized = buffer.feed(deltaContent);
        return {
          ...choice,
          delta: { ...choice.delta, content: detokenized },
        };
      }
      return choice;
    });

    yield { ...chunk, choices };
  }

  // Flush any remaining partial token at stream end.
  const tail = buffer.flush();
  if (tail.length > 0) {
    // Emit a synthetic final chunk carrying the flushed tail on the first choice.
    yield {
      choices: [
        {
          delta: { content: tail },
          index: 0,
          finish_reason: null,
        },
      ],
    } as ChatCompletionChunk;
  }
}

// ---------------------------------------------------------------------------
// Completions proxy
// ---------------------------------------------------------------------------

/**
 * Proxy for `client.chat.completions` that intercepts `create`.
 */
function makeCompletionsProxy(
  realCompletions: {
    create(params: ChatCompletionCreateParamsNonStreaming): Promise<ChatCompletion>;
    create(params: ChatCompletionCreateParamsStreaming): Promise<ChatCompletionStream>;
    create(params: ChatCompletionCreateParams): Promise<ChatCompletion | ChatCompletionStream>;
    [key: string]: unknown;
  },
  pipeline: Pipeline
): typeof realCompletions {
  return new Proxy(realCompletions, {
    get(target, prop: string | symbol) {
      if (prop === "create") {
        return async (
          params: ChatCompletionCreateParams
        ): Promise<ChatCompletion | ChatCompletionStream | AsyncGenerator<ChatCompletionChunk>> => {
          const sessionId = newSessionId();

          // Tokenize messages before sending to OpenAI.
          const tokenizedMessages = await pipeline.tokenizeMessages(
            params.messages,
            sessionId
          );

          if (params.stream === true) {
            // Streaming path
            const streamParams: ChatCompletionCreateParamsStreaming = {
              ...params,
              messages: tokenizedMessages,
              stream: true,
            };
            const rawStream = await (
              target.create as (
                p: ChatCompletionCreateParamsStreaming
              ) => Promise<ChatCompletionStream>
            )(streamParams);

            return detokenizeStream(rawStream, pipeline, sessionId);
          }

          // Non-streaming path
          const nonStreamParams: ChatCompletionCreateParamsNonStreaming = {
            ...params,
            messages: tokenizedMessages,
            stream: false,
          };
          const response = await (
            target.create as (
              p: ChatCompletionCreateParamsNonStreaming
            ) => Promise<ChatCompletion>
          )(nonStreamParams);

          // De-tokenize each choice's message content.
          const detokenizedChoices = response.choices.map((choice) => {
            if (typeof choice.message.content === "string") {
              return {
                ...choice,
                message: {
                  ...choice.message,
                  content: pipeline.detokenize(
                    choice.message.content,
                    sessionId
                  ),
                },
              };
            }
            return choice;
          });

          return { ...response, choices: detokenizedChoices };
        };
      }

      return Reflect.get(target, prop);
    },
  });
}

// ---------------------------------------------------------------------------
// Chat proxy
// ---------------------------------------------------------------------------

function makeChatProxy(
  realChat: { completions: Parameters<typeof makeCompletionsProxy>[0]; [key: string]: unknown },
  pipeline: Pipeline
): typeof realChat {
  return new Proxy(realChat, {
    get(target, prop: string | symbol) {
      if (prop === "completions") {
        return makeCompletionsProxy(target.completions, pipeline);
      }
      return Reflect.get(target, prop);
    },
  });
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/** Minimal shape of an OpenAI client that shieldOpenAI requires. */
export interface OpenAIClientLike {
  chat: {
    completions: {
      create(params: ChatCompletionCreateParamsNonStreaming): Promise<ChatCompletion>;
      create(params: ChatCompletionCreateParamsStreaming): Promise<ChatCompletionStream>;
      create(params: ChatCompletionCreateParams): Promise<ChatCompletion | ChatCompletionStream>;
      [key: string]: unknown;
    };
    [key: string]: unknown;
  };
  [key: string]: unknown;
}

/**
 * Wrap an OpenAI client with PII masking via a Proxy.
 *
 * Intercepts `client.chat.completions.create`, tokenizes messages before
 * sending to OpenAI, and de-tokenizes the response (or each streaming chunk)
 * before returning to the caller.
 *
 * All non-`messages` parameters are forwarded unmodified.
 * No `any` in the public API surface.
 *
 * @param client - An OpenAI (or compatible) client instance.
 * @param pipeline - The PrivacyLens pipeline to use for masking.
 * @returns A Proxy wrapping `client` with PII masking applied.
 */
export function shieldOpenAI<T extends OpenAIClientLike>(
  client: T,
  pipeline: Pipeline
): T {
  return new Proxy(client, {
    get(target, prop: string | symbol) {
      if (prop === "chat") {
        return makeChatProxy(target.chat, pipeline);
      }
      return Reflect.get(target, prop);
    },
  });
}
